semgrep/semgrep/config_resolver.py

```diff
@@ -61,26 +64,178 @@
 }
 
 
+class ConfigPath:
+    _from_registry = False
+    _config_path = ""
+
+    def __init__(self, config_str: str) -> None:
+        """
+        Mutates MetricManager state!
+        Takes a user's inputted config_str and transforms it into the appropriate
+        path, checking whether the config string is a registry url or not. If it
+        is, also set the appropriate MetricManager flag
+        """
+        if config_str in RULES_REGISTRY:
+            self._from_registry = True
+            self._config_path = RULES_REGISTRY[config_str]
+        elif is_url(config_str):
+            self._from_registry = True
+            self._config_path = config_str
+        elif is_registry_id(config_str):
+            self._from_registry = True
+            self._config_path = registry_id_to_url(config_str)
+        elif is_saved_snippet(config_str):
+            self._from_registry = True
+            self._config_path = saved_snippet_to_url(config_str)
+        else:
+            self._from_registry = False
+            self._config_path = config_str
+
+        if self._from_registry:
+            metric_manager.set_using_server_true()
+
+    def resolve_config(self) -> Mapping[str, YamlTree]:
+        """ resolves if config arg is a registry entry, a url, or a file, folder, or loads from defaults if None"""
+        start_t = time.time()
+
+        if self._from_registry:
+            config = self._download_config()
+        else:
+            config = self._load_config_from_local_path()
+
+        if config:
+            logger.debug(f"loaded {len(config)} configs in {time.time() - start_t}")
+        return config
+
+    def _nice_semgrep_url(self, url: str) -> str:
+        """
+        Alters semgrep.dev urls to let user
+        click through to the nice display page instead
+        of raw YAML.
+        Replaces '/c/' in semgrep urls with '/'.
+        """
+        import urllib
+
+        parsed = urllib.parse.urlparse(url)
+        if "semgrep.dev" in parsed.netloc and parsed.path.startswith("/c"):
+            return url.replace("/c/", "/")
+        return url
+
+    def _download_config(self) -> Mapping[str, YamlTree]:
+        config_url = self._config_path
+        downloading_msg = (
+            f"Downloading config from {self._nice_semgrep_url(config_url)}..."
```

Comment:
This will cause URLs to break as they will have `.....` appended to them, c.f.:
<img width="826" alt="Screen Shot 2021-10-04 at 2 47 31 PM" src="https://user-images.githubusercontent.com/409041/135943335-d87ad836-b005-45dc-b0eb-1f201fa8cb84.png">